### PR TITLE
Add History function to kv API

### PR DESCRIPTION
**Pull request by cw9**
Add a History function to get values of past versions

@xichen2020 @robskillington

### CODE REVIEW

File: kv/etcd/store.go [hunk 1]
```diff
@@ -150,17 +155,72 @@ func (c *client) get(key string) (kv.Value, error) {
...
+	if r.Count == 0 {
+		return nil, kv.ErrNotFound
+	}
+
+	numValue := to - from
+
+	latestKV := r.Kvs[0]
+	latestVersion := int(latestKV.Version)
+	latestModRev := latestKV.ModRevision
+	createRev := latestKV.CreateRevision
```
Comment (xichen2020) [inline kv/etcd/store.go]: What's create revision?

File: kv/etcd/store.go [hunk 2]
```diff
@@ -150,17 +155,72 @@ func (c *client) get(key string) (kv.Value, error) {
...
+
+	if latestVersion < from {
+		return nil, errors.New("no value available in the version range")
```
Comment (xichen2020) [inline kv/etcd/store.go]: Maybe just return `nil, nil` here?

File: kv/etcd/store.go [hunk 3]
```diff
@@ -150,17 +155,72 @@ func (c *client) get(key string) (kv.Value, error) {
...
-
 	v := newValue(r.Kvs[0].Value, r.Kvs[0].Version, r.Kvs[0].ModRevision)
 
 	c.mergeCache(key, v)
 
 	return v, nil
 }
 
+func (c *client) History(key string, from, to int) ([]kv.Value, error) {
+	if from >= to || from < 0 || to < 0 {
```
Comment (cw9) [inline kv/etcd/store.go]: sure, can do

File: kv/etcd/store.go [hunk 4]
```diff
@@ -150,17 +155,72 @@ func (c *client) get(key string) (kv.Value, error) {
...
+		return nil, errInvalidHistoryVersion
+	}
+
+	newKey := c.opts.KeyFn()(key)
+
+	ctx, cancel := c.context()
+	r, err := c.kv.Get(ctx, newKey)
+	cancel()
```
Comment (cw9) [inline kv/etcd/store.go]: sure

File: kv/etcd/store.go (hunk 1 above)
Comment (cw9) [inline kv/etcd/store.go]: The store revision when this key is created

File: kv/etcd/store.go (hunk 2 above)
Comment (cw9) [inline kv/etcd/store.go]: sg

File: kv/types.go [hunk 5]
```diff
@@ -95,4 +95,7 @@ type Store interface {
 	// CheckAndSet stores the value for the given key if the current version
 	// matches the provided version
 	CheckAndSet(key string, version int, v proto.Message) (int, error)
+
+	// History returns the value for a key in version range [from, to)
+	History(key string, from, to int) ([]Value, error)
```
Comment (cw9) [inline kv/types.go]: GetBetween seems too vague, GetRange might sound like a function that lists all the keys in a range (which will be added soon)

File: kv/etcd/store.go (hunk 1 above)
Comment (cw9) [inline kv/etcd/store.go]: yeah, I guess I don't really need to check this, will remove

File: kv/etcd/store.go [hunk 6]
```diff
@@ -163,15 +163,16 @@ func (c *client) get(key string) (kv.Value, error) {
 }
 
 func (c *client) History(key string, from, to int) ([]kv.Value, error) {
-	if from >= to || from < 0 || to < 0 {
+	if from > to || from < 0 || to < 0 {
 		return nil, errInvalidHistoryVersion
 	}
```
Comment (xichen2020) [inline kv/etcd/store.go]: As a simple optimization, add `if from == to { return nil, nil }` to avoid unnecessary work

**Pull request merged**